Use more smart pointers in ProcessingInstruction & PseudoElement

cdumez opened the pull request:
#### 728c25250643e662df8f4b9cfa268f62c11bc1b6
<pre>
Use more smart pointers in ProcessingInstruction &amp; PseudoElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=263537">https://bugs.webkit.org/show_bug.cgi?id=263537</a>

Reviewed by Ryosuke Niwa.

* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::protectedContents):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::Document):
(WebCore::Document::checkedStyleScope):
(WebCore::Document::checkedStyleScope const):
* Source/WebCore/dom/Document.h:
(WebCore::Document::styleScope):
(WebCore::Document::styleScope const):
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::~ProcessingInstruction):

(WebCore::ProcessingInstruction::checkStyleSheet):
Drop some unnecessary code that was used to check if we got disconnected
after firing the beforeload event (as you can see in <a href="https://commits.webkit.org/187003@main).">https://commits.webkit.org/187003@main).</a>
We&apos;ve stopped firing the beforeload event a while back but this code wasn&apos;t
simplified.

(WebCore::ProcessingInstruction::isLoading const):
(WebCore::ProcessingInstruction::sheetLoaded):
(WebCore::ProcessingInstruction::setCSSStyleSheet):
(WebCore::ProcessingInstruction::setXSLStyleSheet):
(WebCore::ProcessingInstruction::protectedSheet const):
(WebCore::ProcessingInstruction::parseStyleSheet):
(WebCore::ProcessingInstruction::insertedIntoAncestor):
(WebCore::ProcessingInstruction::removedFromAncestor):
* Source/WebCore/dom/ProcessingInstruction.h:
* Source/WebCore/dom/PseudoElement.cpp:
(WebCore::PseudoElement::create):
(WebCore::PseudoElement::clearHostElement):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::incrementPtrCount const):
(WebCore::Style::Scope::decrementPtrCount const):
(WebCore::Style::Scope::createDocumentResolver):
(WebCore::Style::Scope::releaseMemory):
(WebCore::Style::Scope::didRemovePendingStylesheet):
(WebCore::Style::Scope::addStyleSheetCandidateNode):
(WebCore::Style::Scope::collectActiveStyleSheets):
(WebCore::Style::Scope::updateActiveStyleSheets):
(WebCore::Style::Scope::invalidateStyleAfterStyleSheetChange):
(WebCore::Style::Scope::activeStyleSheetsForInspector):
(WebCore::Style::Scope::flushPendingDescendantUpdates):
(WebCore::Style::Scope::scheduleUpdate):
(WebCore::Style::Scope::collectResolverScopes):
(WebCore::Style::Scope::didChangeStyleSheetEnvironment):
(WebCore::Style::Scope::didChangeViewportSize):
(WebCore::Style::Scope::invalidateMatchedDeclarationsCache):
(WebCore::Style::Scope::pendingUpdateTimerFired):
(WebCore::Style::Scope::documentScope):
(WebCore::Style::Scope::updateQueryContainerState):
* Source/WebCore/style/StyleScope.h:

Canonical link: <a href="https://commits.webkit.org/269694@main">https://commits.webkit.org/269694@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1de84e1fe5159ae4bbcf4c5e51d0e3c62eae1d85

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23268 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1381 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24394 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/25572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21495 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23538 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2902 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23828 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/25572 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23508 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/940 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20153 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/26040 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/753 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21040 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/27202 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21213 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21298 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/752 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18505 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/703 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5558 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1167 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1004 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/ProcessingInstruction.cpp:
```diff
@@ -122,55 +123,42 @@ void ProcessingInstruction::checkStyleSheet()
-            String url = document().completeURL(href).string();
-
-            bool didEventListenerDisconnectThisElement = !isConnected() || &document() != originalDocument.ptr();
-            if (didEventListenerDisconnectThisElement)
```
Why are we dropping this early return?

cdumez replied:
This code used to fire a beforeload event as you can see in https://commits.webkit.org/187003@main.
 That event was dropped but the code wasn't cleaned up. Since no JS runs, anymore, there is no need to check if we got disconnected anymore.

rniwa replied:
Please elaborate that in the commit message.

rniwa commented on Source/WebCore/style/StyleScope.cpp:
```diff
@@ -89,6 +89,16 @@ Scope::~Scope()
 
+void Scope::incrementPtrCount() const
+{
+    m_document->incrementPtrCount();
```
Why are we forwarding it to document?
Style::Scope can be owned by either ShadowRoot or Document.
It doesn't seem correct that we're always forwarding this to Document.

cdumez replied:
Oh, if it can be owned by both then I definitely messed this up.

rniwa commented on Source/WebCore/style/StyleScope.h:
```diff
@@ -74,6 +74,10 @@ class Scope : public CanMakeWeakPtr<Scope> {
     ~Scope();
 
+    // For CheckedPtr / CheckedRef use.
+    void incrementPtrCount() const;
```
We should just make this class inherit from CanMakeCheckedPtr instead unless this is absolutely required in which case the reason must be stated in the commit message.

cdumez replied:
I'll subclass CanMakeCheckedPtr. I missed the ShadowRoot ownership.

cdumez committed:
Use more smart pointers in ProcessingInstruction & PseudoElement
https://bugs.webkit.org/show_bug.cgi?id=263537

Reviewed by Ryosuke Niwa.

* Source/WebCore/css/CSSStyleSheet.cpp:
(WebCore::CSSStyleSheet::protectedContents):
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::Document):
(WebCore::Document::checkedStyleScope):
(WebCore::Document::checkedStyleScope const):
* Source/WebCore/dom/Document.h:
(WebCore::Document::styleScope):
(WebCore::Document::styleScope const):
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::~ProcessingInstruction):

(WebCore::ProcessingInstruction::checkStyleSheet):
Drop some unnecessary code that was used to check if we got disconnected
after firing the beforeload event (as you can see in https://commits.webkit.org/187003@main).
We've stopped firing the beforeload event a while back but this code wasn't
simplified.

(WebCore::ProcessingInstruction::isLoading const):
(WebCore::ProcessingInstruction::sheetLoaded):
(WebCore::ProcessingInstruction::setCSSStyleSheet):
(WebCore::ProcessingInstruction::setXSLStyleSheet):
(WebCore::ProcessingInstruction::protectedSheet const):
(WebCore::ProcessingInstruction::parseStyleSheet):
(WebCore::ProcessingInstruction::insertedIntoAncestor):
(WebCore::ProcessingInstruction::removedFromAncestor):
* Source/WebCore/dom/ProcessingInstruction.h:
* Source/WebCore/dom/PseudoElement.cpp:
(WebCore::PseudoElement::create):
(WebCore::PseudoElement::clearHostElement):
* Source/WebCore/style/StyleScope.cpp:
(WebCore::Style::Scope::incrementPtrCount const):
(WebCore::Style::Scope::decrementPtrCount const):
(WebCore::Style::Scope::createDocumentResolver):
(WebCore::Style::Scope::releaseMemory):
(WebCore::Style::Scope::didRemovePendingStylesheet):
(WebCore::Style::Scope::addStyleSheetCandidateNode):
(WebCore::Style::Scope::collectActiveStyleSheets):
(WebCore::Style::Scope::updateActiveStyleSheets):
(WebCore::Style::Scope::invalidateStyleAfterStyleSheetChange):
(WebCore::Style::Scope::activeStyleSheetsForInspector):
(WebCore::Style::Scope::flushPendingDescendantUpdates):
(WebCore::Style::Scope::scheduleUpdate):
(WebCore::Style::Scope::collectResolverScopes):
(WebCore::Style::Scope::didChangeStyleSheetEnvironment):
(WebCore::Style::Scope::didChangeViewportSize):
(WebCore::Style::Scope::invalidateMatchedDeclarationsCache):
(WebCore::Style::Scope::pendingUpdateTimerFired):
(WebCore::Style::Scope::documentScope):
(WebCore::Style::Scope::updateQueryContainerState):
* Source/WebCore/style/StyleScope.h:

Canonical link: https://commits.webkit.org/269694@main
